### PR TITLE
policy-controller: Honor pod level port annotations

### PR DESCRIPTION
**Pull request by olix0r**
Pods may be annotated with annotations like
`config.linkerd.io/opaque-ports` and
`config.linkerd.io/proxy-require-identity-inbound-ports`--these
annotations configure default behavior that should be honored when a
`Server` does not match the workload's ports. As it stands now, the
policy controller would break opaque-ports configurations that aren't
reflected in a `Server`.

This change reworks the pod indexer to create a default server watch for
each _port_ (rather than for each pod). The cache of default server
watches is now lazy, creating watches as needed for all used
combinations of default policies. These watches are never dropped, but
there are only a few possible combinations of port configurations, so
this doesn't pose any concerns re: memory usage.

While doing this, the names used to describe these default policies are
updated to be prefixed with `default:`. This generally makes these names
more descriptive and easier to understand.

### CODE REVIEW

**Review by adleong — APPROVED** [review]
There are a lot of details around the details of how this is plumbed through the index that I'm still struggling to follow.  But the high level logic seems correct to me.

File: policy-controller/k8s/index/src/defaults.rs [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+
+/// Describes the default behavior for a pod-port to apply when no Server is found for a port.
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq, Hash)]
+pub(crate) struct PortDefaults {
+    pub authenticated: bool,
+    pub opaque: bool,
+}
+
+/// Holds the watches for all default policies. These watches are initiated lazily and are never
+/// updated. `DefaultPolicyWatches` creates instantiating a watch as needed and holding its sender
```
Comment (kleimkuhler) [inline policy-controller/k8s/index/src/defaults.rs:34]:
```suggestion
/// updated. `DefaultPolicyWatches` creates a watch as needed and holding its sender
```

File: policy-controller/k8s/index/src/pod.rs [hunk 2]
```diff
@@ -368,6 +384,78 @@ impl Pod {
...
+            Some(anns) => anns,
+        };
+
+        let opaque = Self::ports_annotation(anns, Self::OPAQUE_PORTS_ANNOTATION);
+        let require_id = Self::ports_annotation(anns, Self::REQUIRE_IDENTITY_PORTS_ANNOTATION);
+
+        Self { opaque, require_id }
+    }
+
+    ///
```
Comment (kleimkuhler) [inline policy-controller/k8s/index/src/pod.rs:406]: remove line?

**Review by kleimkuhler — APPROVED** [review]
Left some trivial comments, but logic and code changes look good.

**Pull request merged**